In-Person Payments: Prevent double-charging of terminal orders

mordeth opened the pull request:
Fixes #6203

#### Changes proposed in this Pull Request

Currently, we determine whether an order was paid purely based on the status of payment intent received in the request. This has resulted in a few cases where payment for an order was charged multiple times with the order's status continuing to be "pending" or "on hold" (ideally, it should transition to "completed" after a successful payment).

This PR adds additional checks based on payment intent information stored in the order's metadata to prevent double-charging.

* Prevent double-charging of paid orders.
* Continue to allow charging of unpaid orders.

Effectively, we are making it impossible to pay for the same order using more than one payment intent.

**Note:** The additional checks are NOT based on the order's status as it can be changed by the user and hence is not as reliable as metadata.

| Intent ID (Order Meta) | Intent Status (Order Meta) | Intent ID (Request) | Allow Capture |
|--------|--------|--------|--------|
| null | null | pi_abc | ✅  |
| pi_abc | requires_capture | pi_abc | ✅  |
| pi_abc | succeeded | pi_abc | ❌  |
| pi_abc | _anything_ | pi_xyz | ❌  |

#### Testing instructions

For all REST API requests below, refer the [docs](https://github.com/woocommerce/woocommerce/wiki/Getting-started-with-the-REST-API) to set up authorization correctly.

##### ✅ Unpaid order (round 1 - using a payment intent with an order for the first time)

* Create a new order in WooCommerce frontend with "Cash on delivery" as the payment method.

* Create a payment intent in Stripe Dashboard. Go to: Your test account → Payments → Create payment → Manual payment. Select the "capture funds later" option.

* On the intent details page, scroll down to the "Metadata" section. Create a new key `order_id` with value from the order created in the previous step.

* Make a POST request to `/wp-json/wc/v3/payments/orders/<order_id>/capture_terminal_payment`.

* Verify that the payment succeeds.

##### ✅ Unpaid order (round 1 - using the same payment intent with a pending/processing/on-hold order) 

* Create a payment intent in Stripe Dashboard. Go to: Your test account → Payments → Create payment → Manual payment. Select the "capture funds later" option.

* Create a new order in WooCommerce frontend with "Cash on delivery" as the payment method. 

* Make a PUT request to `/wp-json/wc/v3/orders/<order_id>`, where order_id refers to the order created in the previous step, to add intent information to the order's metadata.

```json
{
    "meta_data": [
        {
            "key": "_intent_id",
            "value": "<intent id created above>"
        }
    ]
}
```

* On the intent details page , scroll down to the "Metadata" section. Create a new key `order_id` with value from the order created in the previous step.

* Make a POST request to `/wp-json/wc/v3/payments/orders/<order_id>/capture_terminal_payment`.

* Verify that the payment succeeds.

##### ❌ Paid order (round 2 - using a new payment intent with a completed order)

* Create a new order in WooCommerce frontend with "Cash on delivery" as the payment method. 

* Make a PUT request to `/wp-json/wc/v3/orders/<order_id>`, where order_id refers to the order created in the previous step, to add intent information to the order's metadata.

```json
{
    "meta_data": [
        {
            "key": "_intent_id",
            "value": "pi_abc"
        },
        {
            "key": "_intention_status",
            "value": "succeeded"
        },
    ]
}
```

* Create a payment intent in Stripe Dashboard. Go to: Your test account → Payments → Create payment → Manual payment. Select the "capture funds later" option.

* On the intent details page, scroll down to the "Metadata" section. Create a new key `order_id` with value from the order created in the previous step.

* Make a POST request to `/wp-json/wc/v3/payments/orders/<order_id>/capture_terminal_payment`.

* Verify that the payment fails with error code 409.

##### ❌ Unpaid order (round 2 - using a new payment intent with a pending/processing/on-hold order)

* Create a new order in WooCommerce frontend with "Cash on delivery" as the payment method. 

* Make a PUT request to `/wp-json/wc/v3/orders/<order_id>`, where order_id refers to the order created in the previous step, to add intent information to the order's metadata.

```json
{
    "meta_data": [
        {
            "key": "_intent_id",
            "value": "pi_abc"
        }
    ]
}
```

* Create a payment intent in Stripe Dashboard. Go to: Your test account → Payments → Create payment → Manual payment. Select the "capture funds later" option.

* On the intent details page, scroll down to the "Metadata" section. Create a new key `order_id` with value from the order created in the previous step.

* Make a POST request to `/wp-json/wc/v3/payments/orders/<order_id>/capture_terminal_payment`.

* Verify that the payment fails with error code 409.

-------------------

- [ ] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

mordeth committed:
Prevent capture for completed or processed orders

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `6670` or branch name `add/6203-check-for-order-status-on-capture` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=add/6203-check-for-order-status-on-capture) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 5e840f09d18bcd943f3db36214f69b2545436f2c
- Build time: 2023-09-21 07:55:43 UTC

_Note: the build is updated when a new commit is pushed to this PR._

mordeth committed:
Validate intent ID when charging completed order

mordeth committed:
Add unit test for charging completed orders

mordeth committed:
Merge branch 'develop' into add/6203-check-for-order-status-on-capture

mordeth committed:
Add changelog

elazzabi reviewed (approved):
Code looks good 👍 Waiting for testing instructions to test this locally

mordeth committed:
Merge branch 'develop' into add/6203-check-for-order-status-on-capture

dpaun1985 commented:
@mordeth could you please update the testing instructions and mark it as ready for review ?

anu-rock committed:
Simplify code comments

anu-rock committed:
Merge branch 'develop' into add/6203-check-for-order-status-on-capture

anu-rock committed:
Remove order status check

Cash on delivery orders are created with status as "processing".

Furthermore, simply checking for intent in order meta is more reliable.

anu-rock committed:
Add intent status to the check

This will allow an API client to capture payment for an unpaid order using
the payment intent that was previously attached to the order.

anu-rock committed:
Add test case for empty intent status

kalessil commented on includes/admin/class-wc-rest-payments-orders-controller.php:
```diff
@@ -155,6 +155,13 @@ public function capture_terminal_payment( WP_REST_Request $request ) {
+			// Do not process already paid orders to prevent double-charging.
+			$stored_intent_id     = $order->get_meta( WC_Payments_Order_Service::INTENT_ID_META_KEY );
+			$stored_intent_status = $order->get_meta( WC_Payments_Order_Service::INTENTION_STATUS_META_KEY );
+			if ( ! empty( $stored_intent_id ) && ( Intent_Status::SUCCEEDED === $stored_intent_status || $stored_intent_id !== $intent_id ) ) {
```
I'd not make status check dependent on stored intent id and re-worked condition a bit:

`Intent_Status::SUCCEEDED === $stored_intent_status || ($stored_intent_id && $stored_intent_id !== $intent_id)`

kalessil replied:
Can other statuses be stored in the meta? Would it make more reliable perhaps to check for more [statuses](https://stripe.com/docs/api/payment_intents/object#payment_intent_object-status) like processing and cancelled?

anu-rock replied:
It's a valid concern. Attempted to address in ef3d8a9 and fb0ca7b.

kalessil replied:
Thanks!

kalessil commented on tests/unit/admin/test-class-wc-rest-payments-orders-controller.php:
```diff
@@ -1605,6 +1605,116 @@ public function test_get_terminal_intent_payment_method_invalid_value() {
+		$response      = $this->controller->capture_terminal_payment( $request );
+		$response_data = $response->get_data();
+
+		$this->assertEquals( 200, $response->status );
```
Why not strict assertions?

anu-rock replied:
Fixed in bab3118.

kalessil commented on tests/unit/admin/test-class-wc-rest-payments-orders-controller.php:
```diff
@@ -1605,6 +1605,116 @@ public function test_get_terminal_intent_payment_method_invalid_value() {
+			->expects( $this->once() )
+			->method( 'attach_intent_info_to_order' )
+			->with(
+				$this->isInstanceOf( WC_Order::class ),
```
It'd be easier to read if we used `$this->anything()` for parameters which are not essential to verify.

anu-rock replied:
Fixed in bab3118.

anu-rock committed:
Improve typings in unit tests

anu-rock committed:
Include "canceled" intent status in the main condition

anu-rock committed:
Include "processing" intent status in the main condition

kalessil commented on includes/admin/class-wc-rest-payments-orders-controller.php:
```diff
@@ -155,6 +155,20 @@ public function capture_terminal_payment( WP_REST_Request $request ) {
+			$stored_intent_status = $order->get_meta( WC_Payments_Order_Service::INTENTION_STATUS_META_KEY );
+			if (
+				(
+					Intent_Status::SUCCEEDED === $stored_intent_status ||
```
Hint: I'd rather use `in_array` here.

anu-rock replied:
Nice hint. Fixed in c882cb8.

kalessil reviewed (approved):
LGTM (code review only)!

anu-rock committed:
Improve the main condition's readability with in_array

elazzabi reviewed (approved):
Looks good from the code changes 👍 I tried to test it locally but I keep getting some authorizations errors related to WooCommerce REST API (not related with this change)

anu-rock committed:
Merge branch 'develop' into add/6203-check-for-order-status-on-capture

mordeth reviewed (commented):
@anu-rock your code changes are looking good!

anu-rock committed:
Merge branch 'develop' into add/6203-check-for-order-status-on-capture
